### DISCUSSION TITLE
snapshot of dynamic key object

### DISCUSSION THREAD

**Discussion by hyusetiawan** [pmndrs/valtio#622] [Q&A]
I have the following store:
```
export const ColumnSchemaStore = proxy({
  schemas: {} as Record<string, ColumnSchema>,
  // loading: in process to load, error = error loading, done is ready in the schema
  status: {} as { [versionPid: string]: 'loading' | 'error' | 'done' },
})
```
is there a way to subscribe to changes based on a specific property key for schemas? As in:
`useSnapshot(ColumnSchemaStore.schemas[props.schemaId])`
1. Is this possible?
2. What happens if props.schemaId changes?

What i currently end up doing, I get all the schemas and do `schemas[props.schemaId]`. It does work, but I assume, this would get re-rendered whenever there is any change to the schemas object ex: adding/removing property?

**Comment by dai-shi** [comment]
As long as it's not undefined, it should work because if props change, useSnapshot will re-subscribe.

> It does work, but I assume, this would get re-rendered whenever there is any change to the schemas object ex: adding/removing property?

Even if subscribing to whole schemas, rendering will be optimized. So, no worries.

**Comment by dominiczaq** [comment]
Good thing that I've checked all discussions before posting new one 😅 

If object passed to `useSnapshot` is not iterable it will throw an error. But creating subscriptions in multiple components to top level valtio's proxy store may be expensive (initialization, checking for diffs when there is a lot of components with snapshot subscriptions)?

I've created codesandbox with example code https://codesandbox.io/s/valtio-usesnapshot-with-records-49jy5s?file=/src/App.tsx

My question is - shouldn't undefined values for records be handled somehow?

**Reply by dai-shi** [reply]
> My question is - shouldn't undefined values for records be handled somehow?

It's technically impossible to my knowledge, with the current API.

> top level valtio's proxy store may be expensive

Theoretically, yes. But, in terms of React rendering optimization issue, it won't be a bottleneck. I presume if anyone finds a performance drop, it's because `useSnapshot` itself pays some cost. Probably, the difference between `useSnapshot(state)` and `useSnapshot(state.foo)` has much smaller cost than that.

**Reply by dominiczaq** [reply]
Hey, I've used top level subscription for now, and as you said, rendering is still optimised.

I was just thinking if 
```
useSnapshot(store['notExistingKey']);
```
should really throw an error (like it does now) or should it simply return `undefined`.  But looking at the current implementation it would be hard to do this as currently all code assumes that parameter passed to `useSnapshot` is a proxy.

**Reply by dai-shi** [reply]
Theoretically, we could make it undefined, but that would probably break various assumptions, as you said.

**Reply by dominiczaq** [reply]
Yes, as currently it does not seem to be big overhead then  I don't think it's worth pursuing.

Thank you for time and explanations! 🙇

**Reply by dominiczaq** [reply]
@dai-shi sorry to take your time on this one, just wanted to ask before tinkering with the code. Will you be open to accepting the PR if we could make valtio work with undefined values in `useSnapshot` ->`useSnapshot(store['notExistingKey']);`?

**Reply by dai-shi** [reply]
Assuming its complexity, it would be unlikely to be accepted. I'd suggest you start with a third party library, and see how it goes.

In general, our recommendation is, if a key can be non-existent, to destructure a snapshot:
```js
const { notExistingKey } = useSnapshot(store);
```

**Reply by dominiczaq** [reply]
Ok thanks for the heads up. So effectively create new lib which wraps around `useSnapshot` and only rely on public api, does it sounds correct?

**Reply by dai-shi** [reply]
It's nice if you can build it around `useSnapshot`. (Yeah, I think you can do that with fallback proxy.)
Otherwise, creating a new hook with `useSyncExternalStore` is an option, and we can still use valtio/vanilla.
That's basically what I did with [use-valtio](https://github.com/dai-shi/use-valtio).

**Reply by dominiczaq** [reply]
that's really helpful, thanks for the tips and the links 🙇

**Comment by jacob-orbiit** [comment]
I'm trying to do something similar, but with an extra dynamic step:

```js
const product = useSnapshot(state.products[productId]);
```

where

```js
const productsState = new Proxy(proxy({}), {
  get(collection, prop, receiver) {
    if (isProductId(prop)) {
      return fetch(…)
        .then((product) => { productsState[productId] = product });
    }
  },
  set(…) {…},
);
const storeState = proxy({});
const storePromise = new Promise((resolve) => {
  if (!_isEmpty(storeState)) return resolve(storeState);

  gettingStore ??= getStore();

  return gettingStore.then(() => resolve(storeState));
});

export const state = proxy({
  errors,
  store: storePromise,
  products: storePromise.then(() => productsState),
});
```

for the `useSnapshot` above, it doesn't wait for `storePromise` to settle (and it prematurely blows up), but separating them into 2 `useSnapshot`s gets closer:


```js
const { products } = useSnapshot(state);
// by now, `products` has loaded a simple Record<ProductId, Promise<Product>>
const product = useSnapshot(products[productId]);
```

I expected the second `useSnapshot` to wait for that pending promise to settle, but instead it explodes with `undefined is not iterable`.

Setting a breakpoint within the `fetch.then`, I can see that (frustratingly) immediately after valtio explodes, the promise settles and state is updated (but the app has already been unmounted).

I see valtio has logged a warning `Please use proxy object`, but I'm not sure what to do with that—it is a `proxy()`.

P.S. An intermediary (3rd) `useSnapshot(products)` does not help.

**Reply by dai-shi** [reply]
> I see valtio has logged a warning `Please use proxy object`, but I'm not sure what to do with that—it is a `proxy()`.

It means, it's not a valtio proxy.

`new Proxy(proxy({}), {` might be something causing it. Can you somehow avoid it?

**Reply by jacob-orbiit** [reply]
I think no because the `Proxy` needs to intercept the `get` of `state.products[productId]` (which may or may not exist).

**Reply by dai-shi** [reply]
I'm not sure how it's causing the error, maybe if you try to create a minimal reproduction with https://codesandbox.io/s/react-ts, it will reveal it.

**Reply by jacob-orbiit** [reply]
I think this does it: https://codesandbox.io/s/weathered-leftpad-md75zo

**Reply by dai-shi** [reply]
Hm, it looks like it's too complicated. The promise resolution is currently done in `snapshot()` but it expects to return an immutable object (as it can be treated as a snapshot). So, embedding proxies in a snapshot is out of the scope of the usage. I think you may want completely different approach, like using to separate proxy states without embedding.

**Reply by jacob-orbiit** [reply]
What might that look like?

**Reply by dai-shi** [reply]
```js
const state1 = proxy({})
const state2 = proxy({
  promise: something.then((x) => {
    state1.x = x
    return x;
  })
})
```
Not sure if it's a good programming practice.

**Reply by jacob-orbiit** [reply]
How would that work? The keys are not known.

**Reply by dai-shi** [reply]
I don't know the details. Something like this?
```js
const productsState = proxy({})
export const state = proxy({
  products: storePromise.then((data) => {
    const { products } = data;
    return Object.assign(productsState, products);
  }),
})
```
I'm not suggesting this exact thing. I just hope this would open up some new ideas.

**Reply by jacob-orbiit** [reply]
O.o That's a copy-paste of what I put in the CodeSandbox

```js
export const state = proxy({
  products: storePromise.then((data) => {
    const { products } = data;

    return Object.assign(productsState, products);
  }),
  store: storePromise.then((data) => {
    const { products, ...rest } = data;

    return Object.assign(storeState, rest);
  })
});
```

but without the `new Proxy` around productsState's `proxy({})` (and the `Proxy` is needed to intercept the get and set accessors).

**Reply by dai-shi** [reply]
> That's a copy-paste of what I put in the CodeSandbox

Yeah, I thought it would become closer to your mental model.

I didn't mean to return the proxy, which I wanted to avoid:
```js
const productsState = proxy({})
export const state = proxy({
  products: storePromise.then((data) => {
    const { products } = data;
    Object.assign(productsState, products);
    return data;
  }),
})
```

**Reply by jacob-orbiit** [reply]
Okay, so the `Proxy` thing is definitely possible because it suddenly worked (inconsistently):

<img width="1225" alt="image" src="https://user-images.githubusercontent.com/101106010/236480606-f5ce34a1-80fb-407b-9f8e-e62ca59c8fc0.png">

https://codesandbox.io/s/weathered-leftpad-md75zo?file=/src/App.tsx

Break line 10 (eg `useSnapshot(products.p1);` → `useSnapshot(state.products.p1);`) and then "undo" (<kbd>cmd</kbd>+<kbd>z</kbd>), and it works. but refresh, and valtio throws.

**Reply by jacob-orbiit** [reply]
So the problem is that valtio does not like the promise ([`getProduct`](https://codesandbox.io/s/admiring-grass-d2uhnd?file=/src/state.ts:1016-1040)) returned by the `Proxy`: ``Cannot read properties of `undefined` (reading '0')`` `undefined` here is not truly `undefined`, it's just the engine's moronic way of representing an unsettled promise.

@dai-shi valtio ordinarily handles promises; is there a way to get valtio to recognise what is happening and treat it the same? Looking at the `vanilla.js`'s [`snapshot`](https://github.com/pmndrs/valtio/blob/main/src/vanilla.ts#L395), valtio would need to check whether `proxyState` is thenable (which I expected it to already do because under some circumstances, it works).

**Reply by dai-shi** [reply]
maybe we should have a temporary option to avoid unwrapping promise (with wrong types) toward v2. even so, I'm not sure if your use case works. 

> but refresh, and valtio throws.

I think you misunderstand something. It should never work, so if it looks working, it does by chance.

**Reply by jacob-orbiit** [reply]
I think it was "working" because the repro is `proxy(Proxy → Promise<pending>)`, and it was appearing to work because of how CodeSandbox works: The promise settles around the same time as valtio throws (because valtio doesn't detect the thenable); on "refresh", it's not a true refresh, so memory context is persisted and the promise has already settled, so instead of `proxy(Proxy → Promise<pending>)`, it's `proxy(Proxy → value)`.

>maybe we should have a temporary option to avoid unwrapping promise (with wrong types) toward v2.

I'm not sure what this means. But I'm pretty sure if valtio detected the thenable and conditionally handled it async, my case would work just fine. I implemented something similar in nodejs core for ESM loaders (since their functions could, at the time, be async or sync) ([link](https://github.com/nodejs/node/pull/43363/files#diff-f6796082f599554ec3a29c47cf026cb24fc5104884f2632e472c05fe622d778b)).

**Reply by dai-shi** [reply]
> I'm not sure what this means.

Sorry, I was assuming some context. In v2, we will no longer handle promises in our library at all. It doesn't care if a value is a promise or not.

**Reply by JakobJingleheimer** [reply]
Like promises won't work, or this problem is solved?

**Reply by dai-shi** [reply]
Sorry if it wasn't clear. Promises resolving to valtio proxies aren't ever designed usage.

Let me confirm: https://codesandbox.io/s/elastic-ride-m2fx1p?file=/src/App.js
Yeah, we expect a proxy from snapshot, but it's not a proxy.
Hm, that means v2 may fix it (or may not, we don't know yet.)

Wait, does `ref()` workaround? ... No luck. https://codesandbox.io/s/laughing-shockley-14llty?file=/src/App.js
Current implementation does nothing after resolving a promise. (nothing means it doesn't take care `ref()`.)
While this could be fixed in v1, it will break in v2 again.
I think it's better to state this usage is officially unsupported.

Can you modify your use case to avoid storing promises in valtio proxy? and store resolved values in proxy instead?

**Reply by jacob-orbiit** [reply]
> Yeah, we expect a proxy from snapshot, but it's not a proxy.

Does that mean the value passed to `useSnapshot` must be a `proxy`, or that the end-result of what is passed to `useSnapshot` must be a `proxy`? ex

```js
const products = useSnapshot(
  state<ValtioProxy>
    .products<Proxy<Promise<ValtioProxy<Products>>
  );
```
vs

```js
const products = useSnapshot(
  state<ValtioProxy>
    .products<Proxy<Promise<Products>>>
  );
```

> Can you modify your use case to avoid storing promises in valtio proxy? and store resolved values in proxy instead?

I think no? How else will valtio know when to yield the data (which doesn't exist before the promise settles)? [Suspend your components](https://github.com/pmndrs/valtio#suspend-your-components) is in essence, exactly what I want to happen, just slightly more dynamically.

What I'm trying to accomplish is facilitating UI components being blissfully ignorant of whether data is loaded, merely consuming it as if it was, and if it's not, have the state management do whatever is needed to provide it transparently to the UI components. For example:

```js
function ProductsCataloguePage() {
  const { products } = useSnapshot(state);
  // At this point only indexical data (`id` and `name`) is loaded for each item

  return products.map(({ id, name }) => (<Link to={`./${id}`>{name}</Link>));
}
```

```js
function ProductPage() {
  const { productId } = useParams();
  const { name, someAttribute } = useSnapshot(state.products[productId]);
  // Now the rest of this particular product's data is loaded
  // Returning to this page with the same `productId` does not fetch new data;
  // instead, the previous data is served from the data-store (aka valtio).

  return (<h1>{name} <span>{someAttribute}</span></h1>);
}
```

If valtio can't actually do that, then it's unfortunately not a fit (at that point, valtio is basically not doing anything), and I'd rather rip it out now before it gets too intertwined into my app 😞

**Reply by dai-shi** [reply]
I feel very sorry about failing to explain it well.
The last example looks just valid usage.
What's unsupported is to embed a proxy in a promise. In other words, double `useSnapshot` is invalid. (using the return value of useSnapshot to another useSnapshot.)

`<ValtioProxy<Promise<ValtioProxy<Products>>` is unsupported.
`<ValtioProxy<ValtioProxy<Promise<Products>>` is just fine.
`<ValtioProxy<Proxy<ValtioProxy<Products>>` should be fine, but not well tested. So, there might be some issues.

If you feel not a fit, consider something else. Zustand is more predictable.

**Reply by JakobJingleheimer** [reply]
Oh gosh, no need to. It's a complicated and circuitous design.

Thanks! I'll play with that and check out the other lib 🙂 thank you!

**Reply by jacob-orbiit** [reply]
If it turns out to be "necessary", would you be open to me submitting a PR to handle a thenable? I think it would be like <5 lines of code, and a non-breaking change.

**Reply by dai-shi** [reply]
Not sure how you make a change, but

> `<ValtioProxy<Promise<ValtioProxy<Products>>` is unsupported.`

shouldn't be changed. So, changing it wouldn't be acceptable.

btw, if you are familiar with the code base, `unstable_buildProxyFunction` allows you to customize `createSnapshot` so you can try implementing with it.

**Comment by dai-shi** [comment]
Looked at the example: https://codesandbox.io/s/valtio-usesnapshot-with-records-49jy5s?file=/src/App.tsx

https://codesandbox.io/s/valtio-usesnapshot-with-records-49jy5s?file=/src/App.tsx is the canonical solution.

**Reply by jacob-orbiit** [reply]
Is it supposed to crash?
<img width="1217" alt="image" src="https://user-images.githubusercontent.com/101106010/236468450-bb5d8b1e-f4f7-42af-8de5-66dba3a107c8.png">

**Reply by dai-shi** [reply]
Oh, looks like I failed to copy the forked url.
Here's the one I intended: https://codesandbox.io/s/valtio-usesnapshot-with-records-forked-hyhwpv

**Reply by jacob-orbiit** [reply]
Sorry, I'm not sure what I'm supposed to glean from that:

```js
const state = proxy<State>({
  someUser: { message: "this works" }
  // anotherUser: { message: "this does not work if property inside proxy is undefined" }
});

export default function App() {
  const userMessage = useSnapshot(state);
  const anotherUserMessage = userMessage["anotherUser"];
  return (
    <div className="App">
      <h2>someUser message: {userMessage["someUser"]?.message}</h2>
    </div>
  );
}
```

it does not appear to be doing what this discussion is about—nothing there is dynamic.

**Reply by dai-shi** [reply]
I thought `"anotherUser"` is actually dynamic.

Anyway, back to your question:

> My question is - shouldn't undefined values for records be handled somehow?

The answer is no. It always has to be an object.

**Reply by jacob-orbiit** [reply]
I think you're thinking of [dominiczaq](https://github.com/dominiczaq) https://github.com/pmndrs/valtio/discussions/622#discussioncomment-4805368

My question is about getting valtio to simply not blow up—it seems my code _would_ work just fine and valtio would be happy if it just didn't throw prematurely. I think I've just missed some very small piece to get the correct behaviour.

**Reply by dai-shi** [reply]
oops, sorry. then, probably it's not a similar case.

for your case, all I can say is returning a valtio proxy (or any mutable object) from a promise is a bad idea.
hope you find a solution.
